[Investigations] - Add unified components to eql tab

michaelolo24 opened the pull request:
## Summary

This PR introduces the unified table and field list to the correlations tab as seen in the snapshots below.

**As of this PR, items that are not working**
1. Table row height controls
2. Expandable flyout integration
3. Leading cell actions (pinning, notes, row actions, analyzer, session view)

**Changes in this PR:**

Sequence Highlighting:

<img width="1715" alt="image" src="https://github.com/elastic/kibana/assets/17211684/adb63ae3-8709-45b4-8f3b-b84877ac2530">

Building block highlighting:

<img width="1727" alt="Screenshot 2024-04-17 at 1 11 13 PM" src="https://github.com/elastic/kibana/assets/17211684/72f708a9-12c9-4240-a99d-3c9930257c08">


To test:

1. Add `xpack.securitySolution.enableExperimental: [unifiedComponentsInTimelineEnabled]` to your `kibana.dev.yml` 
2. Generate test data (endpoint data is fine)
5. Go to the correlations tab and enter this query to see default events/alerts that should have no highlighting
```any where true```
6. Enter this query to see a generic sequence
```
sequence
[any where true]
[any where true]
```
You can also do something like
```
sequence
[any where host.name=={HOST NAME VALUE HERE}]
[any where true]
```
7. You can also create a correlation rule using any of the above queries to generate building block alerts, and then query those alerts in the correlations tab as well

michaelolo24 committed:
enable unified components in correlations tab

michaelolo24 committed:
add tests and rename hook

janmonschke commented:
@michaelolo24 I noticed that removing a field from the table _feels_ slow:

https://github.com/elastic/kibana/assets/68591/069ae805-87d7-4b09-ab59-4beb8fcd95f4

In comparison, this is how it _feels_ like in Discover:

https://github.com/elastic/kibana/assets/68591/2538663e-febd-4c75-a59d-956c2dd62233

michaelolo24 commented:
> @michaelolo24 I noticed that removing a field from the table _feels_ slow:
> 
>  Screen.Recording.2024-04-17.at.19.41.06.mov 
> In comparison, this is how it _feels_ like in Discover:
> 
>  Screen.Recording.2024-04-17.at.19.56.56.mov

Thanks! Yea, I noticed that as well. I think it has something to do with the row rendering, but maybe I'm just making that up. I noticed the same with the `query` tab with the row renderers, and I believe @logeekal is taking a look

michaelolo24 commented:
/ci

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/eql/index.tsx:
```diff
@@ -161,127 +197,177 @@ export const EqlTabContentComponent: React.FC<Props> = ({
+        </EuiFlexItem>
+      </>
+    ),
+    []
```
The dependencies seem to be missing here

janmonschke replied:
Consider extracting this into its own component

michaelolo24 replied:
Apparently after my Mac updated, my vscode was quarantined and extensions stopped functioning properly. This helped me realize it.. thanks 😅

michaelolo24 replied:
done https://github.com/elastic/kibana/pull/180972/commits/3678fb1dcdbf3aaa0ceea82eb0e83d867164ef7c

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/eql/index.tsx:
```diff
@@ -161,127 +197,177 @@ export const EqlTabContentComponent: React.FC<Props> = ({
+    <EuiFlexGroup gutterSize="s" direction="column">
+      {header}
+    </EuiFlexGroup>
+  ), []);
```
Same here, dependencies seem to be missing

michaelolo24 replied:
see previous comment

michaelolo24 replied:
done https://github.com/elastic/kibana/pull/180972/commits/3678fb1dcdbf3aaa0ceea82eb0e83d867164ef7c

janmonschke commented on x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/use_get_event_type_row_classname.ts:
```diff
@@ -0,0 +1,24 @@
+import type { TimelineItem } from '@kbn/timelines-plugin/common';
+import { useMemo } from "react";
```
Seems like prettier didn't run in this file for some reason 🤔

michaelolo24 replied:
done https://github.com/elastic/kibana/pull/180972/commits/3678fb1dcdbf3aaa0ceea82eb0e83d867164ef7c

PhilippeOberti commented:
When clicking  on the expand button on a row, the flyout isn't opening (works fine without the `unifiedComponentsInTimelineEnabled ` feature flag). At first it seemed related to [this issue](https://github.com/elastic/kibana/issues/180106) but in this PR the flyout just plain does not open, and in the issue it opens but behind timeline...

https://github.com/elastic/kibana/assets/17276605/0ac242f0-e390-40c1-b62b-bb0469c19e7b

PhilippeOberti commented:
I found a behavior that's most likely rarely going to happen but still annoying. I'm not sure we really have control over this though...:
- click on a field in the left sidebar
- click `Visualize`
- click cancel

Expected behavior: we would stay in the same spot we were on

Actual behavior: the view goes back to the `Query` tab


https://github.com/elastic/kibana/assets/17276605/09cd8d43-6d11-4466-8999-894f3749315f

PhilippeOberti commented:
Is the sorting supposed to work? I'm clicking in a bunch of places and nothing seems to be happening...

https://github.com/elastic/kibana/assets/17276605/4e6d0554-a123-42a2-a1a6-d9ef83567a2b

logeekal commented:
> I noticed that removing a field from the table feels slow:

@janmonschke : As @michaelolo24 also pointed out, it is because of combination of `customBodyRender` for rowRenders, `cellActions` , plus some extra updates to redux on our side. I am trying to see how best we can reduce this time. 

Will soon raise a PR.

logeekal reviewed (changes requested):
Thanks for making these changes. Seeing your PR, i remebered, I forgot about building block styling. thanks for getting these done.

Additionally, I see that flyout is not opening in Correlation tab only.


https://github.com/elastic/kibana/assets/7485038/d952a300-1901-42d7-a04f-66a748468115

michaelolo24 committed:
styling pr feedback

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/eql/index.tsx:
```diff
@@ -106,15 +131,26 @@ export const EqlTabContentComponent: React.FC<Props> = ({
     !isEmpty(end) &&
     !isBlankTimeline;
 
+  const defaultColumns = useMemo(
```
Yea, I want to do that for almost all of the tables. The only real difference at this point between them is the query being run..

michaelolo24 replied:
Did a lot of deduping and cleaning up here: https://github.com/elastic/kibana/pull/180972/commits/e37faca8743f8b8d3c007a387fd811c17d584d4e - Let me know if there's anything else you see

logeekal replied:
This looks great. Will check it.

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/custom_timeline_data_grid_body.tsx:
```diff
@@ -117,13 +120,14 @@ const CustomDataGridSingleRow = memo(function CustomDataGridSingleRow(
         : {},
     [canShowRowRenderer]
   );
+  const eventTypeRowClassName = useGetEventTypeRowClassName(rowData.ecs);
```
Just needed to add the className to the `styles.tsx` file here: https://github.com/elastic/kibana/pull/180972/commits/3678fb1dcdbf3aaa0ceea82eb0e83d867164ef7c

michaelolo24 committed:
pr feedback: dedupe code, disable sort buttons, fix flyout

michaelolo24 commented:
@PhilippeOberti 

> When clicking on the expand button on a row, the flyout isn't opening (works fine without the `unifiedComponentsInTimelineEnabled ` feature flag). At first it seemed related to [this issue](https://github.com/elastic/kibana/issues/180106) but in this PR the flyout just plain does not open, and in the issue it opens but behind timeline...

The `query` tab property was hardcoded, updated it [here](https://github.com/elastic/kibana/pull/180972/files#diff-3dd990d9c98493414c158e11b6d1e70b923b7530309a54fff9ffd2f3d36a8b52L152-R174)

> Is the sorting supposed to work? I'm clicking in a bunch of places and nothing seems to be happening...
No, sorting is not supposed to work for the EQL tab as it would mess up the sequence listing for an EQL query. I forgot to check whether the `NO_SORT` would disable the actual buttons themselves, but it didn't, needed to [plumb this through](https://github.com/elastic/kibana/pull/180972/files#diff-3dd990d9c98493414c158e11b6d1e70b923b7530309a54fff9ffd2f3d36a8b52R87): 

FOR: 

> I found a behavior that's most likely rarely going to happen but still annoying. I'm not sure we really have control over this though...:

> Changing the value in the custom Cell row height doesn't seem to be doing anything

I added them to our clean up tasks [here](https://github.com/elastic/security-team/issues/7193)

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/eql/index.tsx:
```diff
@@ -58,6 +61,16 @@ import {
+import { UnifiedTimelineBody } from '../../body/unified_timeline_body';
+import { getColumnHeaders } from '../../body/column_headers/helpers';
+
+const memoizedGetColumnHeaders: (
```
Yep, done

michaelolo24 commented:
> Thanks for making these changes. Seeing your PR, i remebered, I forgot about building block styling. thanks for getting these done.
> 
> Additionally, I see that flyout is not opening in Correlation tab only.

@logeekal see my response to @PhilippeOberti about the same 😄

michaelolo24 committed:
remove unused imports

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/query/header/index.tsx:
```diff
@@ -74,41 +90,69 @@ const QueryTabHeaderComponent: React.FC<Props> = ({
-            iconType="warning"
-            size="s"
-          />
+    <StyledEuiFlyoutHeader data-test-subj={`${activeTab}-tab-flyout-header`} hasBorder={false}>
```
@logeekal , not sure if there was a reason all of this didn't just live here before, but I moved everything for the header here

logeekal replied:
@michaelolo24 , makes sense.. Thanks.

michaelolo24 commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/query/index.tsx:
```diff
@@ -256,79 +217,27 @@ export const QueryTabContentComponent: React.FC<Props> = ({
-        header={header}
-        columns={augumentedColumnHeaders}
+        header={
+          <QueryTabHeader
```
Couldn't decide whether I wanted this behind a memoized const and just pass those two values instead of duplicating them. I can go either way...

logeekal replied:
+1 for having it in `useMemo`. Lot of things going on in timeline and that might recreate header everytime. 

Which values are you talking about?

michaelolo24 committed:
fix file name

PhilippeOberti commented:
Only item that I see missing right now is the fact that despite the fact that have the `expandableTimelineFlyoutEnabled` feature flag in my `kibana.yml` file, I do not see the expandable flyout open when clicking on a row. The old flyout shows up...

PhilippeOberti commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/query/header/index.tsx:
```diff
@@ -74,41 +90,69 @@ const QueryTabHeaderComponent: React.FC<Props> = ({
-        </DataProvidersContainer>
-      ) : null}
-    </EuiFlexGroup>
+        {/* TODO: This is a temporary solution to hide the KPIs until lens components play nicely with timelines */}
+        {/* https://github.com/elastic/kibana/issues/17156 */}
+        {/* <EuiFlexItem grow={false}> */}
+        {/*   <TimelineKpi timelineId={timelineId} /> */}
+        {/* </EuiFlexItem> */}
```
can we please remove this code? It's been sitting there for months (years?) now 😆 

If we ever want to bring it back it will be in the git history...

michaelolo24 replied:
We can, and only like 3 months or so 😂 . Jatin commented it out with the timeline query tab refactor iirc

PhilippeOberti commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/shared/use_timeline_columns.ts:
unit tests for this new hook?

michaelolo24 replied:
yep, will add.

PhilippeOberti commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/eql/header.tsx:
wouldn't we want to unit test this new component?

michaelolo24 replied:
yep, adding! 👍🏾

PhilippeOberti reviewed (commented):
left a few small comments about clean up and adding some unit tests maybe? Biggest thing left I think is [my comment](https://github.com/elastic/kibana/pull/180972#issuecomment-2064790091) related to the expandable flyout not showing up even with the `expandableTimelineFlyoutEnabled ` enabled

michaelolo24 committed:
get expandable flyout to show up

michaelolo24 commented:
> @michaelolo24 only item that I see missing right now is the fact that despite the fact that have the `expandableTimelineFlyoutEnabled` feature flag in my `kibana.yml` file, I do not see the expandable flyout open when clicking on a row. The old flyout shows up...

Yea, the logic for that wasn't here anymore, I added it in this PR: https://github.com/elastic/kibana/pull/180972/commits/8fb51991cf082ac52cbc6cc7ade0bc0a70d29c41

Going to work on adding tests, but just wanted to get this their so you can take a look!

michaelolo24 committed:
remove console log

PhilippeOberti commented:
> > @michaelolo24 only item that I see missing right now is the fact that despite the fact that have the `expandableTimelineFlyoutEnabled` feature flag in my `kibana.yml` file, I do not see the expandable flyout open when clicking on a row. The old flyout shows up...
> 
> Yea, the logic for that wasn't here anymore, I added it in this PR: [8fb5199](https://github.com/elastic/kibana/commit/8fb51991cf082ac52cbc6cc7ade0bc0a70d29c41)
> 
> Going to work on adding tests, but just wanted to get this their so you can take a look!

verified! Thanks for adding this. Soon we'll have the onClose logic working

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

logeekal commented:
>Changing the value in the custom Cell row height doesn't seem to be doing anything

@PhilippeOberti , this behaviour seems to be related to renderCustomBody. I will take this one up.

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/shared/use_timeline_columns.ts:
```diff
@@ -0,0 +1,79 @@
+ */
+
+import { isEmpty } from 'lodash/fp';
+import { useCallback, useMemo } from 'react';
```
Thanks for this abstraction 🥳 .

michaelolo24 replied:
Yea, no worries. But thought about it last night and I should just manage all of this via redux and selectors instead of using this hook. May do a more focused move of custom hooks => useSelector in a tech debt PR I think

logeekal commented:
>only item that I see missing right now is the fact that despite the fact that have the expandableTimelineFlyoutEnabled feature flag in my kibana.yml file, I do not see the expandable flyout open when clicking on a row. The old flyout shows up...

Thank @michaelolo24 for adding new expandable flyout,

Actually after my discussion with @PhilippeOberti , I did the appropriate changes but I kept it aside as a patch as my original PR was already big.

Now that you have already added the code for `expandable flyout`, I am attaching my patch which also includes handling flyout closure. You can either choose to include it or we can also do that it follow up PR.

Only problem right now is, when user closed the flyout, the expand-event icon does not change back. 


[Here is the Patch for Expandable flyout ](https://github.com/elastic/kibana/files/15072338/0001-fix-workaound-for-new-expandable-flyout.patch)

michaelolo24 committed:
tests and revert flyout

logeekal commented:
>I think we should use the default detail expand action the unified data table provides over the one that is in our leadingControlColumns,

Agreed. 💯

>would you mind opening a PR with your patch for testing purposes? 

Will do.

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/styles.tsx:
```diff
@@ -122,11 +124,17 @@ export const StyledTimelineUnifiedDataTable = styled.div.attrs(({ className = ''
+  .udtTimeline .euiDataGridRow:has(.nonRawEvent) {
+    .euiDataGridRowCell--firstColumn,
+    .euiDataGridRowCell--lastColumn {
+      ${({ theme }) => `border-left: 4px solid ${theme.eui.euiColorWarning};`}
```
I think because of this both `building block` and non-buildingblock` events are getting highlighted by a left yellow border. Is this expected?

![grafik](https://github.com/elastic/kibana/assets/7485038/0670d2c2-d0a8-4b35-916e-dfdc01e0e335)

michaelolo24 replied:
Nope, just alerts, not regular events. Regular events have the gray border (that's also how it currently is if you check on main)

logeekal reviewed (commented):
Desk Testing went great 🚀 . Only one small issue as mentioned below. Looking through code now.

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/tabs/shared/use_timeline_columns.ts:
```diff
@@ -0,0 +1,79 @@
+      defaultColumns,
+      getTimelineQueryFieldsFromColumns,
+      leadingControlColumns,
+      localColumns,
```
I am curious what is local columns here?

michaelolo24 replied:
Columns that are passed in to the hook, which are in turn passed to the consuming components. If I remember correctly, it's primarily to allow for external customization of the columns

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/index.tsx:
```diff
@@ -149,27 +151,27 @@ export const TimelineDataTableComponent: React.FC<DataTableProps> = memo(
           timelineActions.toggleDetailPanel({
             ...updatedExpandedDetail,
-            tabType: TimelineTabs.query,
+            tabType: activeTab,
```
Sorry for this screw up 😅  🙏

michaelolo24 replied:
Lol, all good. 😄

logeekal commented on x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/use_get_event_type_row_classname.ts:
```diff
@@ -0,0 +1,34 @@
+        ? isEvenEqlSequence(ecsData)
+          ? 'eqlSequence'
+          : 'eqlNonSequence'
+        : 'nonRawEvent',
```
For my knowledge, what is diff between `nonRawEvent` and `rawEvent`.

logeekal reviewed (approved):
Code looks clean ✨ ✨ . Approving but I have small comments/questions below.

PhilippeOberti commented:
> @PhilippeOberti
> 
> > When clicking on the expand button on a row, the flyout isn't opening (works fine without the `unifiedComponentsInTimelineEnabled ` feature flag). At first it seemed related to [this issue](https://github.com/elastic/kibana/issues/180106) but in this PR the flyout just plain does not open, and in the issue it opens but behind timeline...
> 
> The `query` tab property was hardcoded, updated it [here](https://github.com/elastic/kibana/pull/180972/files#diff-3dd990d9c98493414c158e11b6d1e70b923b7530309a54fff9ffd2f3d36a8b52L152-R174)

@michaelolo24 This is still broken for me. If you don't want to fix it in this PR let's add it to the list in your list of cleanup tasks (link to [my original](https://github.com/elastic/kibana/pull/180972#issuecomment-2062139066) comment for reference)

PhilippeOberti reviewed (approved):
LGTM! It seems that we're tracking everything that still need to be fixed and we'll do follow up PRs.
